lms/services/blackboard_api/client.py

```diff
@@ -186,3 +160,84 @@ def course_groups(
 
         return self_enrollment_groups + instructor_only_groups
+
+    def _list_files_url(self, course_id, folder_id=None) -> str:
+        """Return the url for the list files API endpoint."""
+        path = f"courses/uuid:{course_id}/resources"
+
+        if folder_id:
+            # Get the files and folders in the given folder instead of the
+            # course's top-level files and folders.
+            path += f"/{folder_id}/children"
+
+        return (
+            path
+            + "?"
+            + urlencode(
+                {
+                    "limit": PAGINATION_LIMIT,
+                    "fields": "id,name,type,modified,mimeType,size,parentId",
+                }
+            )
+        )
+
+    def _store_files(self, course_id, files: List[dict]):
+        """Store the given files in the DB."""
+        self._file_service.upsert(
+            [
+                {
+                    "type": "blackboard_file"
+                    if file["type"] == "File"
+                    else "blackboard_folder",
+                    "course_id": course_id,
+                    "lms_id": file["id"],
+                    "name": file["name"],
+                    "size": file["size"],
+                    "parent_lms_id": file["parentId"],
+                }
+                for file in files
+            ]
+        )
+
+    def _get_all_following_pages(self, response, schema) -> List:
+        """
+        Get all the paginated responses to `path`.
+
+        Return the results of all of them in one list after parsing them with `schema`.
+        """
+        results = schema(response).parse()
+        for _ in range(PAGINATION_MAX_REQUESTS - 1):
+            path = response.json().get("paging", {}).get("nextPage")
+            if not path:
+                break
+
+            response = self._api.request("GET", path)
+            results.extend(schema(response).parse())
+
+        return results
+
+    def _get_all_folders_files(self, course_id, results: List[dict]) -> List[dict]:
+        """Get all the files in the folders included in `results` recursively."""
+        # Get all the folder URLs from `results`
+        folder_urls = [
+            self._api._api_url(  # pylint: disable=protected-access
+                self._list_files_url(course_id, result["id"])
+            )
+            for result in results
+            if result["type"] == "Folder"
+        ]
+
+        # Query all folders at once
+        responses = self._request.find_service(name="async_oauth_http").request(
+            "GET", folder_urls
+        )
+
+        for response in responses:
+            # Get all the pages for this folder
+            folder_files = self._get_all_following_pages(
+                response, BlackboardListFilesSchema
+            )
+            # For each folder, get it children too
+            results.extend(self._get_all_folders_files(course_id, folder_files))
+
+        return results
```

Comment:
But this is new I think: getting all the files from all the folders in a course. This looks like it could be a lot of requests!